### DISCUSSION TITLE
mypy plugin: infer types for foreign keys and relationships

### DISCUSSION THREAD

**Discussion by huonw** [sqlalchemy/sqlalchemy#6372] [Ideas]
**Is your feature request related to a problem? Please describe.**

My company is currently happily using SQLAlchemy (thank you!), and mostly happily using https://github.com/dropbox/sqlalchemy-stubs. However, the latter is unmaintained and thus we're regularly hitting issues (and submitting fixes) that don't get fixed or land. Thus, I'd love to switch to the first-party mypy support.

I've tried doing the switch, but found, among other issues, that annotating every foreign key column and relationship property was frustrating.

For instance, previously:

```python
from sqlalchemy import Column, Integer, ForeignKey
from sqlalchemy.orm import relationship
from sqlalchemy.ext.declarative import as_declarative

@as_declarative()
class Base:
    pass

class Foo(Base):
    id = Column(Integer(), primary_key=True)
    bars = relationship("Bar", uselist=True)

class Bar(base):
    id = Column(Integer(), primary_key=True)
    foo_id = Column(ForeignKey(Foo.id))
    foo = relationship("Foo")
```

With sqlalchemy2-stubs:

```python
from sqlalchemy import Column, Integer, ForeignKey
from sqlalchemy.orm import relationship
from sqlalchemy.ext.declarative import as_declarative
from typing import List, Optional

@as_declarative()
class Base:
    pass

class Foo(Base):
    id = Column(Integer(), primary_key=True)
    bars: "Bar" = relationship("Bar", back_populates="foo", uselist=True)

class Bar(Base):
    id = Column(Integer(), primary_key=True)
    foo_id: Optional[int] = Column(ForeignKey(Foo.id))
    foo: Optional[Foo] = relationship("Foo", back_populates="bars")
```

I find this problematic for a few reasons:

- it's error-prone: a mistake in the annotation (like `bars: "Bar"` above: it should be `bars: List["Bar"]`), despite the information presumably being known (e.g. the sqlalchemy-mypy plugin can infer the types correctly)
- it's redundant
- it's more verbose

**Describe the solution you'd like**
It would be great if the mypy plugin could infer more types for these foreign key and relationship properties.

**Describe alternatives you've considered**

Another option would be to require the annotations, but validate that they match the relationship/foreign key targets. This resolves our main concern that the annotations are error prone, but I'm not sure this redundancy is useful (unlike some cases with type annotations).

**Additional context**

I tested the code above using the following versions:

```
greenlet==1.0.0
mypy==0.812
mypy-extensions==0.4.3
SQLAlchemy==1.4.11
sqlalchemy2-stubs==0.0.1a4
typed-ast==1.4.3
typing-extensions==3.7.4.3
```

**Comment by zzzeek** [comment]
>     bars: "Bar" = relationship("Bar", back_populates="foo", uselist=True)

the above will emit an error because you have "uselist=True" and no collection around "Bar".   

> It would be great if the mypy plugin could infer more types for these foreign key and relationship properties.


it's unfortunately (mostly) not possible to resolve object names that are named as strings only in the Mypy plugin, everything else is either implemented or in the case of ForiegnKey(Foo.id) could be figured out at some point.      Mypy plugins work on the source code as a parsed structure and there is no access to any kind of runtime behaviors.   When you have a Column without a type, for the typical ForeignKey() that refers to a string name, there's no way for the plugin to figure out what table / column that's from and the type of that column.   For your example above where it refers to a real Foo.id, *maybe* because we can access Foo.id we can infer the type on that one, but mypy development is super slow and I think we have a lot of much bigger issues still going on.

For relationship to a string, again, we have no idea where "Bar" is in the application, and the rules for resolving those names is [much too complicated to fully reimplement in a plugin](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/clsregistry.py).    It's pretty standard in pep-484 style Python to declare types for things so I think having a few areas in mappings where explicitness is needed for things that are usually runtime inferred is not that big a deal.   People are actually writing their mappings [using dataclasses](https://docs.sqlalchemy.org/en/14/orm/mapping_styles.html#example-two-dataclasses-with-declarative-table) now and even all that verbosity and duplication, people seem to be OK with, so it could definitely be worse.

> 
> Another option would be to require the annotations, but validate that they match the relationship/foreign key targets.

Again this requires that we have any clue where those targets are, which when using strings, is extremely difficult and would probably have a lot more false positives.

> This resolves our main concern that the annotations are error prone, but I'm not sure this redundancy is useful (unlike some cases with type annotations).

The point of the annotations is to reduce errors in the code by running mypy which will tell you where there are problems.  if your annotations are wrong, then mypy will emit errors, or if all the code that uses these annotations are also wrong because they are built around the wrong types, then very basic unit tests would fail.       I think it would be extremely difficult to have the wrong annotations for a relationship that refers to a class by string name, and then have both mypy report no errors as well as have basic unit tests of the code work at the same time.   such an error simply wouldn't be silent on both type-checking and runtime sides.

it's probably more possible for SQLAlchemy to cross-check those annotations *at runtime*, rather than the plugin figuring it out.   that could be done when the mappers configure themselves.   let's think about that.

**Comment by huonw** [comment]
Thanks for the quick reply! It sounds like writing a mypy plugin is hard work, so thanks also for working on it. 👍 

> the above will emit an error because you have "uselist=True" and no collection around "Bar".

Could you be a bit more specific? I purposely included that mistake because it _isn't_ currently flagged by mypy with sqlalchemy2-mypy, AFAICT.

To reproduce what I did, put the code in `test.py`, the config below in `mypy.ini`, install the requirements in the issue, and run `mypy test.py`. 

```ini
[mypy]
plugins = sqlalchemy.ext.mypy.plugin
```

```
$ mypy test.py                             
Success: no issues found in 1 source file
```

> For relationship to a string, again, we have no idea where "Bar" is in the application, and the rules for resolving those names is much too complicated to fully reimplement in a plugin.

`sqlalchemy-stubs` seems to deal with this by requiring models to be imported in order to just resolve the strings within the module. It suggests importing conditionally using `if TYPE_CHECKING:` to avoid (runtime) circularity. What do you think of a solution like this?

> I think it would be extremely difficult to have the wrong annotations for a relationship that refers to a class by string name, and then have both mypy report no errors as well as have basic unit tests of the code work at the same time. such an error simply wouldn't be silent on both type-checking and runtime sides.

Yeah, I'd imagine that in most cases, a mistake here would be caught at during either type checking or tests. I think it's not _unreasonable_ to have models that are similar (e.g. `Person` with `name`, `phone` and others, and `Company` with `name`, `phone` and others) that get mixed up, but potentially not the most important issue to solve.

**Comment by zzzeek** [comment]
> Thanks for the quick reply! It sounds like writing a mypy plugin is hard work, so thanks also for working on it. +1

thanks for that and I hope 1.4 is working out for you.

> 
> > the above will emit an error because you have "uselist=True" and no collection around "Bar".
> 
> Could you be a bit more specific? I purposely included that mistake because it _isn't_ currently flagged by mypy with sqlalchemy2-mypy, AFAICT.

It looks like it's applying that automatically by using List if no collection_cls is present here: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/ext/mypy/infer.py#L142 

so see that, there it *is* helping you and we dont like that either ¯\_(ツ)_/¯

easy enough to put an error there instead, since we know that one.     Please create a distinct issue with this single thing and I'll run it by the mypy people that are helping me out now, who have a lot more opinions about this stuff.

> > For relationship to a string, again, we have no idea where "Bar" is in the application, and the rules for resolving those names is much too complicated to fully reimplement in a plugin.
> 
> `sqlalchemy-stubs` seems to deal with this by requiring models to be imported in order to just resolve the strings within the module. It suggests importing conditionally using `if TYPE_CHECKING:` to avoid (runtime) circularity. What do you think of a solution like this?

OK I see what they're doing there, and sure if you want it that relationship("NAME"") means we look right in the current module for NAME and that's it, otherwise we ask for an explicit annotation, that's easy enough.   I guess when placing an explicit annotation you are also having to do the import at the top in any case, so yeah, that makes sense now, thanks for explaining that.   

It's not as easy for a foreign key with ("Column.id") as a string, since we'd have to eval that expression so I'd rather not get into that.  also it wont work for something like relationship("foo.bar.NAME") unless we decide some means of guessing how to interpret that (just use the rightmost token?), but you never see anyone using that syntax anyway.   again if you can add a single issue as a "usecase" and indicate relationship("NAME") would link to an immedaite symbol of that name in the application we can add that.   because sure we are making you do that anyway, so yeah.


> 
> > I think it would be extremely difficult to have the wrong annotations for a relationship that refers to a class by string name, and then have both mypy report no errors as well as have basic unit tests of the code work at the same time. such an error simply wouldn't be silent on both type-checking and runtime sides.
> 
> Yeah, I'd imagine that in most cases, a mistake here would be caught at during either type checking or tests. I think it's not _unreasonable_ to have models that are similar (e.g. `Person` with `name`, `phone` and others, and `Company` with `name`, `phone` and others) that get mixed up, but potentially not the most important issue to solve.

if the TYPE_CHECKING thing with "immedaite name match" is good enough then we are probably OK with that.   it's just i couldnt try to open up the whole application and look for "Bar".

**Reply by huonw** [reply]
> thanks for that and I hope 1.4 is working out for you.

Yeah, it's great 😄 

> It looks like it's applying that automatically by using List if no collection_cls is present here: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/ext/mypy/infer.py#L142
>
> so see that, there it is helping you and we dont like that either ¯_(ツ)_/¯

Ah, I see, I hadn't noticed that there's inference for non-string `relationship` arguments 👍 If I flip the code around in order to pass the class value itself in, I get the sort of error I'd expect: works well in that case!

```
test.py:20: error: [SQLAlchemy Mypy plugin] Left hand assignment 'bars: "Bar"' not compatible with ORM mapped expression of type "Mapped[List[Bar]]"
```

<details> 

```python
from sqlalchemy import Column, Integer, ForeignKey
from sqlalchemy.orm import relationship
from sqlalchemy.ext.declarative import as_declarative
from typing import List, Optional


@as_declarative()
class Base:
    pass


class Bar(Base):
    id = Column(Integer(), primary_key=True)
    foo_id: Optional[int] = Column(ForeignKey("foo.id"))
    foo: Optional["Foo"] = relationship("Foo", back_populates="bars", uselist=False)


class Foo(Base):
    id = Column(Integer(), primary_key=True)
    bars: Bar = relationship(Bar, back_populates="foo", uselist=True)
```

</details>

> easy enough to put an error there instead, since we know that one. Please create a distinct issue with this single thing and I'll run it by the mypy people that are helping me out now, who have a lot more opinions about this stuff.

I think my concerns would be resolved by making string arguments work in more cases, and so I haven't filed an issue for this specifically yet.

> OK I see what they're doing there, and sure if you want it that relationship("NAME"") means we look right in the current module for NAME and that's it, otherwise we ask for an explicit annotation, that's easy enough. I guess when placing an explicit annotation you are also having to do the import at the top in any case, so yeah, that makes sense now, thanks for explaining that.

> again if you can add a single issue as a "usecase" and indicate relationship("NAME") would link to an immedaite symbol of that name in the application we can add that. because sure we are making you do that anyway, so yeah.

No worries, it seems we're both learning a lot from this discussion 👍 

https://github.com/sqlalchemy/sqlalchemy/issues/6407 (including a brief discussion of a small implementation I tried).
